### PR TITLE
Improvements in test logs

### PR DESCRIPTION
**Pull request by narrieta**
* Created individual logs for the setup of the test environments (previously they were written to the LISA log)
* Split the logs in directories: tests, environment setup, lisa, and vm_logs
* Removed duplicate junit test results for test failures (it used to show on the test results and on the LISA results)

### CODE REVIEW

File: tests_e2e/orchestrator/lib/agent_test_suite.py [hunk 1]
```diff
@@ -49,7 +50,7 @@
 from tests_e2e.tests.lib.agent_test import TestSkipped
 from tests_e2e.tests.lib.agent_test_context import AgentTestContext
 from tests_e2e.tests.lib.identifiers import VmIdentifier
-from tests_e2e.tests.lib.logging import log as agent_test_logger  # Logger used by the tests
```
Comment (narrieta) [inline tests_e2e/orchestrator/lib/agent_test_suite.py:52]: Initally most of the output was going to the LISA log, which was referenced as 'self.log', while the the test logs were referenced as 'agent_test_logger'.

Over time, most of the output has been moved to the test logs. Now the LISA log is references as 'self.lisa_log' and the test logs simply as 'log' (which makes this class consistent with the naming in the rest of the test code).

File: tests_e2e/orchestrator/lib/agent_test_suite.py [hunk 2]
```diff
@@ -114,10 +115,10 @@ def __init__(self, vm: VmIdentifier, paths: AgentTestContext.Paths, connection:
             super().__init__(vm=vm, paths=paths, connection=connection)
             # These are initialized by AgentTestSuite._set_context().
             self.log_path: Path = None
-            self.log: Logger = None
+            self.lisa_log: Logger = None
             self.node: Node = None
             self.runbook_name: str = None
-            self.image_name: str = None
```
Comment (narrieta) [inline tests_e2e/orchestrator/lib/agent_test_suite.py:120]: the value in this variable is actually the environment name; renaming accordingly

File: tests_e2e/orchestrator/lib/agent_test_suite.py [hunk 3]
```diff
@@ -129,12 +130,10 @@ def __init__(self, metadata: TestSuiteMetadata) -> None:
         # The context is initialized by _set_context() via the call to execute()
         self.__context: AgentTestSuite._Context = None
 
-    def _set_context(self, node: Node, variables: Dict[str, Any], lisa_log_path: str, log: Logger):
+    def _initialize(self, node: Node, variables: Dict[str, Any], lisa_working_path: str, lisa_log_path: str, lisa_log: Logger):
```
Comment (narrieta) [inline tests_e2e/orchestrator/lib/agent_test_suite.py:133]: renamed the method to make explicit that this initializes the AgentTestSuite

File: tests_e2e/orchestrator/lib/agent_test_suite.py [hunk 4]
```diff
@@ -129,12 +130,10 @@ def __init__(self, metadata: TestSuiteMetadata) -> None:
...
         connection_info = node.connection_info
         node_context = get_node_context(node)
         runbook = node.capability.get_extended_runbook(AzureNodeSchema, AZURE)
-        # Remove the resource group and node suffix, e.g. "e1-n0" in "lisa-20230110-162242-963-e1-n0"
-        runbook_name = re.sub(r"-\w+-\w+$", "", runbook.name)
```
Comment (narrieta) [inline tests_e2e/orchestrator/lib/agent_test_suite.py:137]: this was use to create a unique directory name for the runbook; no longer needed

File: tests_e2e/orchestrator/lib/agent_test_suite.py [hunk 5]
```diff
@@ -214,20 +219,22 @@ def _setup(self) -> None:
         self._setup_lock.acquire()
 
         try:
-            self._log.info("")
-            self._log.info("**************************************** [Build] ****************************************")
-            self._log.info("")
+            log.info("")
```
Comment (narrieta) [inline tests_e2e/orchestrator/lib/agent_test_suite.py:222]: The setup of the environment now has its own log

File: tests_e2e/orchestrator/lib/agent_test_suite.py [hunk 6]
```diff
@@ -298,113 +308,136 @@ def _setup_node(self) -> None:
...
+
+    def _execute(self, environment: Environment, variables: Dict[str, Any]):
         """
         Executes each of the AgentTests included in the "c_test_suites" variable (which is generated by the AgentTestSuitesCombinator).
         """
-        self._set_context(node, variables, log_path, log)
+        # Set the thread name to the name of the environment. The thread name is added to each item in LISA's log.
+        with _set_thread_name(self.context.environment_name):
+            log_path: Path = self.context.log_path/f"env-{self.context.environment_name}.log"
+            with set_current_thread_log(log_path):
```
Comment (narrieta) [inline tests_e2e/orchestrator/lib/agent_test_suite.py:373]: each environment now has its own log

File: tests_e2e/orchestrator/lib/agent_test_suite.py [hunk 7]
```diff
@@ -488,47 +524,67 @@ def _check_agent_log(self) -> bool:
...
+                log.info("Detected %s error(s) in the agent log:\n\n%s", len(errors), '\n'.join(['\t' + e.text for e in errors]))
+            self._mark_log_as_failed(log_path)
+
+            self._report_test_result(
+                self.context.environment_name,
+                "CheckAgentLog",
+                TestStatus.FAILED,
+                start_time,
+                message=message + '\n' + '\n'.join([e.text for e in errors[0:3]]))
+        except:    # pylint: disable=bare-except
```
Comment (narrieta) [inline tests_e2e/orchestrator/lib/agent_test_suite.py:569]: we were not handling exceptions raised while collecting the vm logs; now we are

File: tests_e2e/orchestrator/lib/agent_test_suite.py [hunk 8]
```diff
@@ -488,47 +524,67 @@ def _check_agent_log(self) -> bool:
...
+                self.context.environment_name,
+                "CheckAgentLog",
+                TestStatus.FAILED,
+                start_time,
+                "Error checking agent log",
+                add_exception_stack_trace=True)
 
         return False
 
+    @staticmethod
```
Comment (narrieta) [inline tests_e2e/orchestrator/lib/agent_test_suite.py:581]: Logs for test suites with errors are now prefixed with '_'

File: tests_e2e/pipeline/scripts/execute_tests.sh [hunk 9]
```diff
@@ -70,29 +79,33 @@ docker run --rm \
...
 #
-# Remove the 2 levels of the tree that indicate the time of the test run to make navigation
-# in the Azure Pipelines UI easier. Also, move the lisa log one level up and remove some of
-# the logs that are not needed
-#
-mv "$BUILD_ARTIFACTSTAGINGDIRECTORY"/lisa/[0-9][0-9][0-9][0-9][0-9][0-9][0-9][0-9]/*/* "$BUILD_ARTIFACTSTAGINGDIRECTORY"/lisa
-rm -r "$BUILD_ARTIFACTSTAGINGDIRECTORY"/lisa/[0-9][0-9][0-9][0-9][0-9][0-9][0-9][0-9]
-mv "$BUILD_ARTIFACTSTAGINGDIRECTORY"/lisa/lisa-*.log "$BUILD_ARTIFACTSTAGINGDIRECTORY"
-rm "$BUILD_ARTIFACTSTAGINGDIRECTORY"/lisa/messages.log
+if ls "$LOGS_DIRECTORY"/env-*.log > /dev/null 2>&1; then
```
Comment (narrieta) [inline tests_e2e/pipeline/scripts/execute_tests.sh:89]: separated the logs into subdirectories:
* root directory: logs for tests with errors
* test_logs: logs for passing tests
* environment_logs: logs for environment (test VM) setup
* vm_logs: logs harvested from the VMs
* runbook_logs: logs global to the runbook (LISA log and junit results)

File: tests_e2e/tests/lib/logging.py [hunk 10]
```diff
@@ -127,8 +145,11 @@ def set_current_thread_log(log_file: Path):
     """
     Context Manager to set the log file for the current thread temporarily
     """
+    initial_value = log.get_current_thread_log()
```
Comment (narrieta) [inline tests_e2e/tests/lib/logging.py:148]: This allows nesting context managers

**Review by maddieford — APPROVED** [review]

File: tests_e2e/orchestrator/lib/agent_test_suite.py (hunk 6 above)
Comment (nagworld9) [reply tests_e2e/orchestrator/lib/agent_test_suite.py:373]: sorry I'm really confused here, how many different sets of logs we have? 

For example
```

        self.context.lisa_log.info("Setting up test node")
        log.info("")
```
The first log goes to main log which is gist of each step of the flow or only logs specific msgs?
The second is what you are calling environment log which has something like node step, installing test version logs etc..?

where about test scenario log goes? will it have separate log?

Comment (narrieta) [reply tests_e2e/orchestrator/lib/agent_test_suite.py:373]: @nagworld9 I'd recommend checking the artifacts in the test run to see each log and its contents

Comment (nagworld9) [reply tests_e2e/orchestrator/lib/agent_test_suite.py:373]: makes sense. 
Basically log.info could be either test log or environment log depending on which thread under its being invoked.

File: tests_e2e/tests/bvts/vm_access.py [hunk 11]
```diff
@@ -51,7 +52,7 @@ def run(self):
         private_key_file: Path = self._context.working_directory/f"{username}_rsa"
         public_key_file: Path = self._context.working_directory/f"{username}_rsa.pub"
         log.info("Generating SSH key as %s", private_key_file)
-        ssh: SshClient = SshClient(ip_address=self._context.vm_ip_address, username=username, private_key_file=private_key_file)
+        ssh = SshClient(ip_address=self._context.vm_ip_address, username=username, private_key_file=private_key_file)
```
Comment (nagworld9) [inline tests_e2e/tests/bvts/vm_access.py:55]: why type removed?

File: tests_e2e/orchestrator/lib/agent_test_suite.py [hunk 12]
```diff
@@ -298,113 +308,136 @@ def _setup_node(self) -> None:
...
+            with set_current_thread_log(log_path):
+                start_time: datetime.datetime = datetime.datetime.now()
+                success = True
+
+                try:
+                    # Log the environment's name and the variables received from the runbook (note that we need to expand the names of the test suites)
+                    log.info("LISA Environment (for correlation with the LISA log): %s", environment.name)
+                    log.info("Runbook variables:")
+                    for name, value in variables.items():
+                        log.info("    %s: %s", name, value if name != 'c_test_suites' else [t.name for t in value])
```
Comment (nagworld9) [inline tests_e2e/orchestrator/lib/agent_test_suite.py:382]: is extra space at the start intentional?

File: tests_e2e/orchestrator/lib/agent_test_suite.py (hunk 6 above)
Comment (narrieta) [reply tests_e2e/orchestrator/lib/agent_test_suite.py:373]: Not quite.

We have multiple threads executing multiple environments, each environment executing multiple test suites. For each environment, we have an environment log and multiple test suite logs. If we had 3 environments executing 5 test suites each, we would have 3 environment logs + 3x5 test suite logs (total 18 logs, plus some other logs: the LISA log, the CheckAgentLog log, etc).

For each thread we have this nesting

```
log_path: Path = self.context.log_path/f"env-{self.context.environment_name}.log"
with set_current_thread_log(log_path):
    ....
    log.info("GOES TO THE ENVIRONMENT LOG")
    ...  
    for suite in self.context.test_suites:  # pylint: disable=E1133
    ....
        log_path:Path = self.context.log_path/f"{suite_full_name}.log"
        with set_current_thread_log(log_path):
             ....
            log.info("GOES TO THE TEST SUITE LOG")
    ....
    log.info("GOES TO THE ENVIRONMENT LOG AGAIN")

```

Comment (narrieta) [reply tests_e2e/orchestrator/lib/agent_test_suite.py:373]: i.e the scope of the log is determined by the context manager in the with statement

File: tests_e2e/tests/bvts/vm_access.py (hunk 11 above)
Comment (narrieta) [reply tests_e2e/tests/bvts/vm_access.py:55]: it's already declared on line 41

File: tests_e2e/orchestrator/lib/agent_test_suite.py (hunk 12 above)
Comment (narrieta) [reply tests_e2e/orchestrator/lib/agent_test_suite.py:382]: yes, it adds a level of indentation that helps with readability

**Review by nagworld9 — APPROVED** [review]

**Pull request merged**